feat: Implement categories feature API integration

rajeh1032 opened the pull request:
<!--
  🛠️ For Work In Progress, please use Draft PRs:
  https://github.blog/2019-02-14-introducing-draft-pull-requests/

  ✅ Before submitting:
  - Follow your project's contributing guide
  - Ensure code compiles and passes all tests
  - Use descriptive commit messages
  - Keep PRs small and focused
-->

## 📌 What type of PR is this? (check all applicable)

- [  ✅] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🧹 Refactor
- [ ] 🚀 Performance Optimization
- [ ] 🧾 Documentation Update
- [ ] ♿ Accessibility Fix
- [ ] 🔧 CI/CD or Config Change

---

## 📝 Description

_Describe the goal of this PR and what was changed._
- Added API service for fetching all categories using Dio and Retrofit.
- Created remote data source implementation for categories.
- Developed mappers for converting DTOs to domain entities.
- Established repository pattern for categories data management.
- Implemented use case for retrieving categories.
- Built view model to handle categories state and events.
- Added unit tests for remote data source and repository implementations.

---

## 🧩 Related Issues / Tickets

- Closes #
- Fixes #
- Related to #

---

## 🧪 QA Instructions

_How to test the changes manually?_

- [ ✅ ] Run the app on Android
- [ ] Run the app on iOS
- [ ] Check specific feature/page:
  - [ ] ...
- [ ] Verify responsiveness on different screen sizes

---

## 📸 Screenshots or Recordings (if UI changes)

_Add screenshots or screen recordings to demonstrate your changes._
<img width="1082" height="1248" alt="image" src="https://github.com/user-attachments/assets/03863100-7265-462e-88aa-a8084468222d" />
<img width="598" height="1359" alt="image" src="https://github.com/user-attachments/assets/f0d77683-4006-472e-b1e0-843d5227fa37" />



---

## ♿ UI Accessibility Checklist (if applicable)

- [ ] Semantic widget structure used (`Semantics`, `Accessible`, etc.)
- [ ] Supports screen readers
- [ ] Good color contrast
- [ ] Tested with TalkBack/VoiceOver

---

## 🧪 Tests

- [ ✅ ] Unit tests added
- [ ] Widget tests added
- [ ] Integration tests added
- [ ] No tests needed (explain why):

---

## 🛠 Post-Deployment Tasks (Optional)

_Any migrations, scripts, or setup tasks?_

---

## 🎬 Fun GIF (Optional)

_What GIF best represents this PR?_

![fun gif](https://media.giphy.com/media/xT0xeJpnrWC4XWblEk/giphy.gif)

rajeh1032 committed:
feat: Implement categories feature API integration

- Added API service for fetching all categories using Dio and Retrofit.
- Created remote data source implementation for categories.
- Developed mappers for converting DTOs to domain entities.
- Established repository pattern for categories data management.
- Implemented use case for retrieving categories.
- Built view model to handle categories state and events.
- Designed UI for displaying categories with error handling.
- Added unit tests for remote data source and repository implementations.
- Created storage interface for future data persistence needs.

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/models/metadata_dto.dart:
```diff
@@ -0,0 +1,29 @@
+import 'package:json_annotation/json_annotation.dart';
+
```
remove spaces in all files ....

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/models/response/category_response_dto.dart:
```diff
@@ -0,0 +1,27 @@
+import 'package:flower_e_commerce_app/Feature/mainLayout/tabs/categoriesFeature/api/models/category_dto.dart';
+import 'package:flower_e_commerce_app/Feature/mainLayout/tabs/categoriesFeature/api/models/metadata_dto.dart';
+import 'package:json_annotation/json_annotation.dart';
+
```
space

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/models/category_dto.dart:
```diff
@@ -0,0 +1,46 @@
+import 'package:json_annotation/json_annotation.dart';
+
```
space

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/mapper/category_dto_mapper .dart:
```diff
@@ -0,0 +1,15 @@
+        name: name ?? "",
+        slug: slug ?? "",
+        image: image ?? "",
+        createdAt: DateTime.now(),
```
remove from here un wanted parameters

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/domain/entities/category_entity.dart:
```diff
@@ -0,0 +1,21 @@
+  final String name;
+  final String slug;
+  final String image;
+  final DateTime createdAt;
```
we don't need this date times into our entity we will not show the user any date this info not helpful to the user at any cases  remove them from here and remove him from mapper to be entity less noise

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/domain/entities/category_entity.dart:
```diff
@@ -0,0 +1,21 @@
+  final String image;
+  final DateTime createdAt;
+  final DateTime updatedAt;
+  final bool isSuperAdmin;
```
we dont need this to be in entity

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/domain/entities/category_entity.dart:
```diff
@@ -0,0 +1,21 @@
+  final DateTime createdAt;
+  final DateTime updatedAt;
+  final bool isSuperAdmin;
+  final int productsCount;
```
we dont need this to be in entity

Eng-M0ataz commented on lib/Feature/auth/api/client/api_service.dart:
why you upload this ?

Eng-M0ataz commented on devtools_options.yaml:
dont upload this

Eng-M0ataz commented on assets/translations/en.json:
dont upload this

Eng-M0ataz commented on android/build/reports/problems/problems-report.html:
dont upload this

Eng-M0ataz commented on assets/translations/ar.json:
dont upload this

rajeh1032 replied:
what about string that i will added?

rajeh1032 committed:
feat: Implement category products fetching functionality

- Added API service methods to fetch products by category.
- Created request and response models for category products.
- Implemented mappers for converting DTOs to domain entities.
- Updated repository and data source to handle product fetching.
- Enhanced ViewModel to manage product state and events.

Eng-M0ataz commented on .vscode/settings.json:
why i see this

Eng-M0ataz commented on assets/translations/ar.json:
what is thhis ????

Eng-M0ataz commented on assets/translations/en.json:
what is thhis ????

Eng-M0ataz commented on devtools_options.yaml:
why you upload this ?

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/mapper/product_dto_mapper.dart:
map only the data you need in screen

Eng-M0ataz commented on lib/Feature/auth/api/client/api_service.dart:
what is this ???

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/mapper/product_response_dto_mapper.dart:
```diff
@@ -0,0 +1,12 @@
+extension ProductResponseDtoMapper on ProductResponseDto {
+  ProductResponseEntity toEntity() => ProductResponseEntity(
+        message: message ?? "",
+        metadata: metadata?.toEntity(),
```
handle nullable value

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/models/product_dto.dart:
```diff
@@ -0,0 +1,82 @@
+import 'package:json_annotation/json_annotation.dart';
+
```
remove spaces

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/domain/entities/product_entity.dart:
remove all unwanted parm

Eng-M0ataz commented on lib/core/Widgets/custom_app_bar.dart:
why you changed this ?

rajeh1032 replied:
this branch from Development and this files does not updates

Eng-M0ataz commented on lib/core/utils/Constants/app_routes.dart:
```diff
@@ -3,4 +3,5 @@ abstract class AppRoutes {
   static const String signInRoute = '/loginRoute';
   static const String signUpRoute = '/registerRoute';
   static const String forgetPasswordRoute = '/forgetPasswordRoute';
+  static const String categoriesRoute = '/categoryScreenRoute';
```
'/categoryScreenRoute  why you used screen in string ? its should be '/categoryRoute

rajeh1032 replied:
for test i will remove this route bro the nav bar will open screen

Eng-M0ataz commented on pubspec.lock:
dont upload tis

rajeh1032 committed:
feat,refacture: Implement category tabs, , unit test,Api Call  All Products End point , Ui of Category feature

- Added `CategoryTabs` widget for displaying categories with selection functionality.
- Introduced `FilterButton` for triggering filter actions.
- Created `FilterIconButtonOfAppBar` for filter access in the app bar.
- Developed `ProductCard` for displaying product details.
- Implemented `ProductGrid` for organizing products in a grid layout.
- Added `SearchBarWidget` for searching products.
- Included shimmer effects for loading states in `CategoryTabs` and `ProductCard`.
- Created unit tests for `CategoriesViewModel` to ensure proper state management and API interactions.
- include Unit test

Eng-M0ataz commented on assets/translations/ar.json:
```diff
@@ -1 +1,10 @@
+  "no_products": "'لا يوجد منتجات متاحة'",
+  "add_to_cart": "'اضافة للسلة'",
+  "no_categories": "'لا يوجد تصنيفات متاحة'",
+  "all": "All",
```
this is an English word why it's here in Arabic file?

Eng-M0ataz commented on assets/translations/ar.json:
```diff
@@ -1 +1,10 @@
-{}
\ No newline at end of file
+{
+  "sign_up": "إنشاء حساب",
```
its a categories feature why this line is here ?

Eng-M0ataz commented on assets/translations/en.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "sign_up": "Sign up",
```
same here

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/client/api_service.dart:
```diff
@@ -0,0 +1,27 @@
+  Future<CategoryResponseDto> getAllCategories();
+
+  @GET(ApiConstants.getProducts)
+  Future<ProductResponseDto> getProducts(
```
this methode you named it in data source  getCategoryProducts  and here you  named it getProducts so use the same name in all files and rename it to be better like gitProductsByCategory.

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/dataSources/categories_remote_data_source_impl.dart:
```diff
@@ -0,0 +1,67 @@
+    try {
+      final requestModel = GetCategoryProductsRequestModel.fromEntity(
+          getCategoryProductsRequestEntity);
+      var response = await _apiServices.getProducts(requestModel);
```
see here the name is not the same like the method

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/models/request/get_category_products_request_model.dart:
```diff
@@ -0,0 +1,30 @@
+import 'package:flower_e_commerce_app/Feature/mainLayout/tabs/categoriesFeature/domain/entities/requestEntities/get_category_products_request_entity.dart';
+
+class GetCategoryProductsRequestModel {
```
rename getProductsByCategoryInputModel

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/models/request/get_category_products_request_model.dart:
```diff
@@ -0,0 +1,30 @@
+  Map<String, dynamic> toJson() {
+    return {
+      "category": categoryId,
+      if (page != null) "page": page,
```
"page": page ?? null,

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/models/request/get_category_products_request_model.dart:
```diff
@@ -0,0 +1,30 @@
+    this.limit,
+  });
+
+  factory GetCategoryProductsRequestModel.fromEntity(
```
we dont need an entity to this model this model only send  parm  with req

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/data/dataSources/categories_remote_data_source.dart:
```diff
@@ -0,0 +1,13 @@
+import 'package:flower_e_commerce_app/Feature/mainLayout/tabs/categoriesFeature/domain/entities/responseEntities/product_response_entity.dart';
+import 'package:flower_e_commerce_app/core/Errors/api_results.dart';
+
+abstract interface class CategoriesRemoteDataSource {
```
rename the methods to match name in api serveice

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/data/repositories/categories_repo_impl.dart:
rename the methods here to

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/domain/repositories/categories_repo.dart:
rename methods here to

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/domain/useCases/get_category_products_use_case.dart:
```diff
@@ -0,0 +1,20 @@
+  GetCategoryProductsUseCase({required CategoriesRepo categoriesRepo})
+      : _categoriesRepo = categoriesRepo;
+  Future<ApiResult<ProductResponseEntity>> invoke(
+      {required String? categoryId, required int? page, required int? limit}) {
```
why you dont use model here you created and used parm

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/domain/useCases/get_category_products_use_case.dart:
```diff
@@ -0,0 +1,20 @@
+  Future<ApiResult<ProductResponseEntity>> invoke(
+      {required String? categoryId, required int? page, required int? limit}) {
+    return _categoriesRepo.getCategoryProducts(GetCategoryProductsRequestEntity(
+      categoryId: categoryId ?? '',
```
we don't handle null here you already handle it in model

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/pages/category_screen.dart:
```diff
@@ -0,0 +1,19 @@
+    return BlocProvider<CategoriesViewModel>(
+      create: (context) =>
+          getIt<CategoriesViewModel>()..doIntent(GetAllCategoriesEvent()),
+      child: const CategoryScreenBody(),
```
return here all screen needs like scaffold and app bar and safe area here the main body its only have the body of the screen

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/events/categories_event.dart:
```diff
@@ -0,0 +1,31 @@
+import 'package:equatable/equatable.dart';
+
+sealed class CategoriesEvent extends Equatable {
```
we dont need to extend Equatable  here we only use it in state so we can make unite test

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/events/categories_event.dart:
```diff
@@ -0,0 +1,31 @@
+}
+
+class GetAllCategoriesEvent extends CategoriesEvent {
+  const GetAllCategoriesEvent();
```
and we dont need the constracor here

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/events/categories_event.dart:
```diff
@@ -0,0 +1,31 @@
+  });
+
+  @override
+  List<Object?> get props => [categoryId, page, limit];
```
remove this

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/events/categories_event.dart:
```diff
@@ -0,0 +1,31 @@
+  const GetAllProductsEvent();
+}
+
+class GetCategoryProductsEvent extends CategoriesEvent {
```
we dont need to accept anything here

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/states/categories_state.dart:
```diff
@@ -0,0 +1,49 @@
+
+class CategoriesState extends Equatable {
+  final bool isLoading;
+  final String? errorMessage;
```
accept the failure here not error message and in the ui access  failure.errorMessage and in failure you  have the code to  if you need to show it

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/states/categories_state.dart:
```diff
@@ -0,0 +1,49 @@
+class CategoriesState extends Equatable {
+  final bool isLoading;
+  final String? errorMessage;
+  final List<CategoryEntity>? categories;
```
why you make them null here you already in parsing checks if the list return with null you return an empty  list

Eng-M0ataz commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/viewModel/states/categories_state.dart:
```diff
@@ -0,0 +1,49 @@
+  final String? selectedCategoryId;
+
+  const CategoriesState({
+    this.isLoading = false,
```
you can make this true because  the user when  enter the screen the loading already show and this will make you save emit in the first

ali-safwat commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/models/request/get_category_products_request_model.dart:
```diff
@@ -0,0 +1,30 @@
+
+class GetCategoryProductsRequestModel {
+  final String categoryId;
+  final int? page;
```
request model all feilds required

remove nullable

ali-safwat commented on lib/Feature/mainLayout/tabs/categoriesFeature/api/dataSources/categories_remote_data_source_impl.dart:
```diff
@@ -0,0 +1,67 @@
+
+  @override
+  Future<ApiResult<CategoryResponseEntity>> getAllCategories() async {
+    try {
```
use excute api call func to avoid repeating try catch

ali-safwat commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/pages/category_screen.dart:
```diff
@@ -0,0 +1,19 @@
+  Widget build(BuildContext context) {
+    return BlocProvider<CategoriesViewModel>(
+      create: (context) =>
+          getIt<CategoriesViewModel>()..doIntent(GetAllCategoriesEvent()),
```
avoid use getIt inside widgit build method

ali-safwat commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/widgets/category_tabs.dart:
```diff
@@ -0,0 +1,103 @@
+          if (index == 0) {
+            final isSelected = selectedCategoryId == 'all';
+            return _buildCategoryTab(
+                categoryId: 'all',
```
hard coded

ali-safwat commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/widgets/category_tabs.dart:
```diff
@@ -0,0 +1,103 @@
+        itemCount: categories.length + 1,
+        itemBuilder: (context, index) {
+          if (index == 0) {
+            final isSelected = selectedCategoryId == 'all';
```
hard coded

ali-safwat commented on lib/Feature/mainLayout/tabs/categoriesFeature/presentation/widgets/product_card.dart:
```diff
@@ -0,0 +1,197 @@
+  }
+
+  Widget _buildPriceRow(BuildContext context) {
+    final priceAfterDiscount =
```
move it to view model

ali-safwat reviewed (changes requested):
request changes

john-safwat reviewed (commented):
pr is too large

rajeh1032 commented:
Feature was large and i couldn't separated it in many PR Because Code
Review


‫في الأربعاء، 3 سبتمبر 2025 في 12:40 ص تمت كتابة ما يلي بواسطة ‪John
Safwat‬‏ ***@***.***‬‏>:‬

> ***@***.**** commented on this pull request.
>
> pr is too large
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Eng-M0ataz/Flower-E-Commerce-App/pull/6#pullrequestreview-3178211087>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXOPGH47XJZV7LXFOR34UFT3QYFFXAVCNFSM6AAAAACEWMP4GKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCNZYGIYTCMBYG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

rajeh1032 committed:
refactor: replace request entity with model and improve code consistency

- Replace GetCategoryProductsRequestEntity with GetCategoryProductsRequestModel
- Update all references to use the new model throughout the codebase
- Add executeApi utility function for consistent error handling
- Improve code formatting and add const constructors where applicable
- Add app constants for better maintainability
- Update imports and remove unused entity file

rajeh1032 committed:
Refactor categories feature to use consistent naming and improve UI theming

- Rename GetCategoryProductsUseCase to GetProductsByCategoryUseCase for consistency
- Rename getCategoryProducts to getProductsByCategory across data layer
- Update request model class name to GetProductsByCategoryRequestModel
- Replace hardcoded colors with theme-based colors in UI components
- Add shimmer color constants and sizing constants for better maintainability
- Fix null safety issues in entity classes by making fields nullable
- Add executeApi utility function for consistent API error handling
- Update test files to reflect renamed classes and methods
- Improve category tabs with null-safe property access

rajeh1032 committed:
Replace shimmer package with shimmer_animation and update theming

rajeh1032 committed:
fix
